app/src/main/java/net/aiscope/gdd_app/ui/sample_completion/SampleCompletionViewModel.kt

```diff
@@ -0,0 +1,179 @@
+package net.aiscope.gdd_app.ui.sample_completion
+
+import android.content.Context
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import net.aiscope.gdd_app.R
+import net.aiscope.gdd_app.model.BloodQuality
+import net.aiscope.gdd_app.model.CompletedCapture
+import net.aiscope.gdd_app.model.MicroscopeQuality
+import net.aiscope.gdd_app.model.SampleMetadata
+import net.aiscope.gdd_app.model.SamplePreparation
+import net.aiscope.gdd_app.model.SampleStatus
+import net.aiscope.gdd_app.model.WaterType
+import net.aiscope.gdd_app.network.RemoteStorage
+import net.aiscope.gdd_app.repository.SampleRepository
+import net.aiscope.gdd_app.ui.metadata.FieldOption
+import net.aiscope.gdd_app.ui.metadata.MetadataMapper
+import timber.log.Timber
+import javax.inject.Inject
+
+class SampleCompletionViewModel @Inject constructor(
+    private val repository: SampleRepository,
+    private val remoteStorage: RemoteStorage,
+    private val context: Context,
+) : ViewModel() {
+    //TODO: We got a nicer way of fixing default vals??
+
+    // Fields for the metadata tab
+    var disease: String = ""
+    var captures: List<CompletedCapture> = emptyList()
+    var options: List<FieldOption> = emptyList()
+    var required: Boolean = true
+    var smearTypeId: Int? = null
+    var speciesValue: String? = null
+    var comments: String? = null
+
+    // Fields for preparation tab. Perhaps we should nest this?
+    var waterType: String = context.getString(R.string.spinner_empty_option)
+    var usesGiemsa: Boolean = true
+    var giemsaFP: Boolean = true
+    var usesPbs: Boolean = true
+    var reusesSlides: Boolean = false
+    var bloodQuality: String = context.getString(R.string.spinner_empty_option)
+
+    //Fields for the microscope tab
+    var microscopeDamaged: Boolean = false
+    var microscopeMagnification: Int = 1000
+
+
+    override fun toString(): String {
+        val builder = StringBuilder()
+        this.javaClass.declaredFields.forEach {
+            builder.append(
+                "* " + it.name + " : " + this.javaClass.getDeclaredField(it.name).get(this) + "\n"
+            )
+        }
+        return builder.toString()
+    }
+
+    fun initVM() {
+        //TODO: inject dispatchers
+        viewModelScope.launch(Dispatchers.IO) {
+
+            val sample = repository.current()
+            disease = sample.disease
+            captures = sample.captures.completedCaptures
+
+            //So what we gonna do here for the meta stuff?
+            val lastMeta = repository.lastSaved()?.metadata
+            Timber.i("LAST META %s", lastMeta)
+            smearTypeId = MetadataMapper.getSmearTypeId(lastMeta?.smearType!!)

```

Comment:
Hello, it's better not to use !! as much as posible because if it's null the app will crash. You have two options, or control the value with a if( != null) or a let?. or another option is to Modify the mappers to accept a null input, and control in the mapper that in the case of null set a default value.

Reply:
Changed the mapper to take nullable types